src/qibojit/custom_operators/backends.py

```diff
@@ -384,3 +383,252 @@ def swap_pieces(self, piece0, piece1, new_global, nlocal):
     def measure_frequencies(self, frequencies, probs, nshots, nqubits, seed=1234, nthreads=None):
         raise NotImplementedError("`measure_frequencies` method is not "
                                   "implemented for GPU.")
+
+
+class CuQuantumBackend(CupyBackend): # pragma: no cover
+    # CI does not test for GPU
+
+    def __init__(self):
+        super(CuQuantumBackend, self).__init__()
+        import cuquantum # pylint: disable=import-error
+        from cuquantum import custatevec as cusv # pylint: disable=import-error
+        self.cuquantum = cuquantum 
+        self.cusv = cusv
+        self.name = "cuquantum"
+
+    def get_cuda_type(self, dtype='complex64'):
+        if dtype == 'complex128':
+            return self.cuquantum.cudaDataType.CUDA_C_64F, self.cuquantum.ComputeType.COMPUTE_64F
+        elif dtype == 'complex64':
+            return self.cuquantum.cudaDataType.CUDA_C_32F, self.cuquantum.ComputeType.COMPUTE_32F
+        else:
+            raise TypeError("Type can be either complex64 or complex128")
+
+    def one_qubit_base(self, state, nqubits, target, kernel, qubits=None, gate=None):
+        # initialize cuStateVec library
+        handle = self.cusv.create()
+
+        ntarget = 1
+        target = nqubits - target - 1
+        target = self.np.asarray([target], dtype = self.np.int32)
+        ncontrols = len(qubits) - 1 if qubits is not None else 0
+        controls = self.np.asarray([i for i in qubits.get() if i != target], dtype = self.np.int32)
+        adjoint = 0
+        # implement x, y, z and z_pow kernel since cuquantum only has apply_matrix
+        # TODO: find a better way to implement this
+        if kernel == "apply_z_pow":
+            u1gate = self.cp.zeros((2, 2),dtype=state.dtype)
+            u1gate[0, 0], u1gate[1, 1] = 1, gate
+            gate = u1gate
```

Comment:
For the ZPow we could change the `apply_z_pow` in `JITCustomBackend` to:
```Python
def apply_z_pow(self, state, gate, nqubits, targets, qubits=None):
    if self.engine.name == "cuquantum":
        # calculate 2x2 gate
    return self.engine.one_qubit_base(state, nqubits, *targets, "apply_z_pow", qubits, gate)
```
The same would work for fSim below. 

With this and the above comment we should avoid all if conditions in the cuquantum methods.

Reply:
Ok, sure also for the SWAP gate right?